storage/fcds/fcds.go

```diff
@@ -0,0 +1,333 @@
+}
+
+// NewStore constructs a new Store with files at path, with specified max chunk size.
+// Argument withCache enables in memory cache of free chunk data positions in files.
+func NewStore(path string, maxChunkSize int, metaStore MetaStore, withCache bool) (s *Store, err error) {
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return nil, err
+	}
+	shards := make(map[byte]*os.File, shardCount)
+	shardsMu := make(map[uint8]*sync.Mutex)
+	for i := byte(0); i < shardCount; i++ {
+		shards[i], err = os.OpenFile(filepath.Join(path, fmt.Sprintf("chunks-%v.db", i)), os.O_CREATE|os.O_RDWR, 0666)
+		if err != nil {
+			return nil, err
+		}
+		shardsMu[i] = new(sync.Mutex)
+	}
+	var freeCache *offsetCache
+	if withCache {
+		freeCache = newOffsetCache(shardCount)
+	}
+	return &Store{
+		shards:       shards,
+		shardsMu:     shardsMu,
+		meta:         metaStore,
+		freeCache:    freeCache,
+		free:         make(map[uint8]struct{}),
+		maxChunkSize: maxChunkSize,
+		quit:         make(chan struct{}),
+	}, nil
+}
+
+// Get returns a chunk with data.
+func (s *Store) Get(addr chunk.Address) (ch chunk.Chunk, err error) {
+	done, err := s.protect()
+	if err != nil {
+		return nil, err
+	}
+	defer done()
+
+	mu := s.shardsMu[getShard(addr)]
+	mu.Lock()
+	defer mu.Unlock()
+
+	m, err := s.getMeta(addr)
+	if err != nil {
+		return nil, err
+	}
+	data := make([]byte, m.Size)
+	n, err := s.shards[getShard(addr)].ReadAt(data, m.Offset)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	if n != int(m.Size) {
+		return nil, fmt.Errorf("incomplete chunk data, read %v of %v", n, m.Size)
+	}
+	return chunk.NewChunk(addr, data), nil
+}
+
+// Has returns true if chunk is stored.
+func (s *Store) Has(addr chunk.Address) (yes bool, err error) {
+	done, err := s.protect()
+	if err != nil {
+		return false, err
+	}
+	defer done()
+
+	mu := s.shardsMu[getShard(addr)]
+	mu.Lock()
+	defer mu.Unlock()
+
+	_, err = s.getMeta(addr)
+	if err != nil {
+		if err == chunk.ErrChunkNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Put stores chunk data.
+func (s *Store) Put(ch chunk.Chunk) (err error) {
+	done, err := s.protect()
+	if err != nil {
+		return err
+	}
+	defer done()
+
+	addr := ch.Address()
+	shard := getShard(addr)
+	f := s.shards[shard]
+	data := ch.Data()
+	section := make([]byte, s.maxChunkSize)
+	copy(section, data)
+
+	s.freeMu.RLock()
+	_, hasFree := s.free[shard]
+	s.freeMu.RUnlock()
+
+	var offset int64
+	var reclaimed bool
+	mu := s.shardsMu[shard]
+	mu.Lock()
+	if hasFree {
+		var freeOffset int64 = -1
+		if s.freeCache != nil {
+			freeOffset = s.freeCache.get(shard)
+		}
+		if freeOffset < 0 {
+			freeOffset, err = s.meta.FreeOffset(shard)
+			if err != nil {
+				return err
+			}
+		}
+		if freeOffset < 0 {
+			offset, err = f.Seek(0, io.SeekEnd)
+			if err != nil {
+				mu.Unlock()
+				return err
+			}
+			s.freeMu.Lock()
+			delete(s.free, shard)
```

Comment:
> why set this here and not in FreeOfffset func?

FreeOffset function is a function meta, this delete is on internal Store field.

Reply:
> use bool array here? or `*int32` array and use atomic set/load without freeMu needed

There is a data race condition with atomic operations setting uint32 bitvector. When a bitvector is set, first its value has to be loaded, then a specific bit set and then stored again. Because the stored integer value is derived from the loaded one, all three operations has to be locked, which is not possible with atomic functions, there has to be a lock to protect the stored value while the new one is computed. If a value is loaded and stored with atomic functions it is possible that another goroutine can change the integer value and the last store would overwrite with its, now wrong value.

Given that, I think that a slice of booleans and a lock are most suitable, as you suggested in other comment.

Reply:
I am not saying bitvector. I am saying that you can alternatively use a fix array of int pointers (i think int32 is the smallest that atomic allows) and set and read with atomic on each position instead of locking the entire array. This allows more parallelism as well as slightly faster

Reply:
Store.free is marking only if a shard has any offsets. I have changed it to use slice of booleans.